src/function/scalar/map/map_from_entries.cpp

```diff
@@ -0,0 +1,85 @@
+#include "duckdb/planner/expression/bound_function_expression.hpp"
+#include "duckdb/common/string_util.hpp"
+#include "duckdb/parser/expression/bound_expression.hpp"
+#include "duckdb/function/scalar/nested_functions.hpp"
+#include "duckdb/common/types/data_chunk.hpp"
+#include "duckdb/common/pair.hpp"
+
+namespace duckdb {
+
+static void MapFromEntriesFunction(DataChunk &args, ExpressionState &state, Vector &result) {
+	D_ASSERT(result.GetType().id() == LogicalTypeId::MAP);
+
+	// ! Otherwise if its not a constant vector, this breaks the optimizer
+	result.SetVectorType(VectorType::CONSTANT_VECTOR);
+	for (idx_t i = 0; i < args.ColumnCount(); i++) {
+		if (args.data[i].GetVectorType() != VectorType::CONSTANT_VECTOR) {
+			result.SetVectorType(VectorType::FLAT_VECTOR);
+		}
+	}
+
+	// Get the arguments vector
+	auto &array = args.data[0];
+	idx_t pair_amount = ListVector::GetListSize(array);
+	auto &entries = ListVector::GetEntry(array);
+
+	// Prepare the result vectors
+	result.Resize(0, pair_amount);
+	auto &child_entries = StructVector::GetEntries(result);
+	D_ASSERT(child_entries.size() == 2);
+	auto &key_vector = child_entries[0];
+	auto &value_vector = child_entries[1];
+
+	auto key_data = FlatVector::GetData<list_entry_t>(*key_vector);
+	auto value_data = FlatVector::GetData<list_entry_t>(*value_vector);
+	// Transform to mapped values
+	key_data[0].offset = 0;
+	value_data[0].offset = 0;
+	for (idx_t i = 0; i < pair_amount; i++) {
+		auto element = entries.GetValue(i);
+		D_ASSERT(element.type().id() == LogicalTypeId::STRUCT);
+		auto &key_value = StructValue::GetChildren(element);
+		ListVector::PushBack(*key_vector, key_value[0]);
+		ListVector::PushBack(*value_vector, key_value[1]);
+	}
+	key_data[0].length = pair_amount;
+	value_data[0].length = pair_amount;
+
+	result.Verify(args.size());
+}
+
+static unique_ptr<FunctionData> MapFromEntriesBind(ClientContext &context, ScalarFunction &bound_function,
+                                                   vector<unique_ptr<Expression>> &arguments) {
+	child_list_t<LogicalType> child_types;
+
+	if (arguments.size() != 1) {
+		throw Exception("We need one list of structs for a map");
+	}
+	auto &list = arguments[0]->return_type;
```

Comment:
I dont know enough about the usecases of maps, or this function, to determine if it should ever be NULL, or just throw an error instead.
I'll inquire on this before making a decision +1

Reply:
Yes, sounds good. Try to find out the desired behaviour for `SELECT map_from_entries(NULL)`. I think the (for me) logical result of this would be `NULL`.

Reply:
Spend way too long figuring out how spark-shell works, but never the less:
- any element is NULL
```
scala> spark.sql("SELECT map_from_entries(array(struct(1, 'b'), null))").toDF().show()
+-------------------------------------------+
|map_from_entries(array(struct(1, b), NULL))|
+-------------------------------------------+
|                                       null|
+-------------------------------------------+
```
- 0 arguments is a no-go, it errors
- null as argument (instead of a list)
```
scala> spark.sql("SELECT map_from_entries(null)").toDF().show()
org.apache.spark.sql.AnalysisException: cannot resolve 'map_from_entries(NULL)' due to data type mismatch: 'NULL' is of void type. map_from_entries accepts only arrays of pair structs.; line 1 pos 7;
```
- null as map key
```
scala> spark.sql("SELECT map_from_entries(array(struct(null, 'b')))").toDF().show()
java.lang.RuntimeException: Cannot use null as map key!
```
non-unique keys
```
scala> spark.sql("SELECT map_from_entries(array(struct(1, null), struct(1, 'b')))").toDF().show()
java.lang.RuntimeException: Duplicate map key 1 was found, please check the input data. If you want to remove the duplicated keys, you can set spark.sql.mapKeyDedupPolicy to LAST_WIN so that the key inserted at last takes precedence.
```
- null as value is perfectly fine